:Polars


In [10]:
import polars as pl
from google.colab import files
uploaded = files.upload()

Saving train.csv to train.csv


In [27]:
df = pl.read_csv("train.csv")

In [12]:
print(df.head())

shape: (5, 12)
┌─────────────┬──────────┬────────┬───────────────────┬───┬───────────┬─────────┬───────┬──────────┐
│ PassengerId ┆ Survived ┆ Pclass ┆ Name              ┆ … ┆ Ticket    ┆ Fare    ┆ Cabin ┆ Embarked │
│ ---         ┆ ---      ┆ ---    ┆ ---               ┆   ┆ ---       ┆ ---     ┆ ---   ┆ ---      │
│ i64         ┆ i64      ┆ i64    ┆ str               ┆   ┆ str       ┆ f64     ┆ str   ┆ str      │
╞═════════════╪══════════╪════════╪═══════════════════╪═══╪═══════════╪═════════╪═══════╪══════════╡
│ 1           ┆ 0        ┆ 3      ┆ Braund, Mr. Owen  ┆ … ┆ A/5 21171 ┆ 7.25    ┆ null  ┆ S        │
│             ┆          ┆        ┆ Harris            ┆   ┆           ┆         ┆       ┆          │
│ 2           ┆ 1        ┆ 1      ┆ Cumings, Mrs.     ┆ … ┆ PC 17599  ┆ 71.2833 ┆ C85   ┆ C        │
│             ┆          ┆        ┆ John Bradley (Fl… ┆   ┆           ┆         ┆       ┆          │
│ 3           ┆ 1        ┆ 3      ┆ Heikkinen, Miss.  ┆ … ┆ STON/O2.  ┆ 7.92

In [13]:
df. describe() #описательная статистика

statistic,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
str,f64,f64,f64,str,str,f64,f64,f64,str,f64,str,str
"""count""",891.0,891.0,891.0,"""891""","""891""",714.0,891.0,891.0,"""891""",891.0,"""204""","""889"""
"""null_count""",0.0,0.0,0.0,"""0""","""0""",177.0,0.0,0.0,"""0""",0.0,"""687""","""2"""
"""mean""",446.0,0.383838,2.308642,null,null,29.699118,0.523008,0.381594,null,32.204208,null,null
"""std""",257.353842,0.486592,0.836071,null,null,14.526497,1.102743,0.806057,null,49.693429,null,null
"""min""",1.0,0.0,1.0,"""Abbing, Mr. Anthony""","""female""",0.42,0.0,0.0,"""110152""",0.0,"""A10""","""C"""
"""25%""",224.0,0.0,2.0,null,null,20.0,0.0,0.0,null,7.925,null,null
"""50%""",446.0,0.0,3.0,null,null,28.0,0.0,0.0,null,14.4542,null,null
"""75%""",669.0,1.0,3.0,null,null,38.0,1.0,0.0,null,31.0,null,null
"""max""",891.0,1.0,3.0,"""van Melkebeke, Mr. Philemon""","""male""",80.0,8.0,6.0,"""WE/P 5735""",512.3292,"""T""","""S"""


In [14]:
df. dtypes #типы данных

[Int64,
 Int64,
 Int64,
 String,
 String,
 Float64,
 Int64,
 Int64,
 String,
 Float64,
 String,
 String]

In [15]:
df.null_count() #количество пропусков

PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
u32,u32,u32,u32,u32,u32,u32,u32,u32,u32,u32,u32
0,0,0,0,0,177,0,0,0,0,687,2


In [16]:
df.mean() #среднее значение для числовых столбцов

PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
f64,f64,f64,str,str,f64,f64,f64,str,f64,str,str
446.0,0.383838,2.308642,null,null,29.699118,0.523008,0.381594,null,32.204208,null,null


In [17]:
counts = df["Pclass"].value_counts() #количество пассажиров каждого класса (Pclass)
print(counts)

shape: (3, 2)
┌────────┬───────┐
│ Pclass ┆ count │
│ ---    ┆ ---   │
│ i64    ┆ u32   │
╞════════╪═══════╡
│ 1      ┆ 216   │
│ 2      ┆ 184   │
│ 3      ┆ 491   │
└────────┴───────┘


In [18]:
result = (
    df
    .group_by("Sex")
    .agg(
        pl.col("Survived").sum().alias("survived_count")
    )
)

print(result) #количество выживших мужчин и женщин на корабле

shape: (2, 2)
┌────────┬────────────────┐
│ Sex    ┆ survived_count │
│ ---    ┆ ---            │
│ str    ┆ i64            │
╞════════╪════════════════╡
│ male   ┆ 109            │
│ female ┆ 233            │
└────────┴────────────────┘


In [19]:
filtered_df = df.filter(pl.col("Age") > 44) #часть таблицы с пассажирами, возраст которых больше 44 лет

print(filtered_df)

shape: (115, 12)
┌─────────────┬──────────┬────────┬──────────────────┬───┬────────────┬─────────┬───────┬──────────┐
│ PassengerId ┆ Survived ┆ Pclass ┆ Name             ┆ … ┆ Ticket     ┆ Fare    ┆ Cabin ┆ Embarked │
│ ---         ┆ ---      ┆ ---    ┆ ---              ┆   ┆ ---        ┆ ---     ┆ ---   ┆ ---      │
│ i64         ┆ i64      ┆ i64    ┆ str              ┆   ┆ str        ┆ f64     ┆ str   ┆ str      │
╞═════════════╪══════════╪════════╪══════════════════╪═══╪════════════╪═════════╪═══════╪══════════╡
│ 7           ┆ 0        ┆ 1      ┆ McCarthy, Mr.    ┆ … ┆ 17463      ┆ 51.8625 ┆ E46   ┆ S        │
│             ┆          ┆        ┆ Timothy J        ┆   ┆            ┆         ┆       ┆          │
│ 12          ┆ 1        ┆ 1      ┆ Bonnell, Miss.   ┆ … ┆ 113783     ┆ 26.55   ┆ C103  ┆ S        │
│             ┆          ┆        ┆ Elizabeth        ┆   ┆            ┆         ┆       ┆          │
│ 16          ┆ 1        ┆ 2      ┆ Hewlett, Mrs.    ┆ … ┆ 248706     ┆ 16

Ускорение работы с pandas

In [20]:
import pandas as pd
import bottleneck as bn
import numpy as np


In [21]:
df_pd = pd.read_csv("train.csv")

In [22]:
age_np = df["Age"].to_numpy()

mean_age = bn.nanmean(age_np)
std_age = bn.nanstd(age_np, ddof=1)

print(f"Средний возраст: {mean_age:.2f}")
print(f"Стандартное отклонение: {std_age:.2f}")

Средний возраст: 29.70
Стандартное отклонение: 14.53


In [23]:
df_pd["Fare_new"] = df_pd["Fare"].apply(lambda x: x * 1.3) #умножаем значение столбца Fare на 1.3

Оптимизация типов pandas

In [24]:
uploaded = files.upload()

Saving Housing.csv to Housing (4).csv


In [25]:
df_housing = pd.read_csv("Housing.csv")
df_housing.head()

,price,area,bedrooms,bathrooms,stories,mainroad,guestroom,basement,hotwaterheating,airconditioning,parking,prefarea,furnishingstatus
0,13300000,7420,4,2,3,yes,no,no,no,yes,2,yes,furnished
1,12250000,8960,4,4,4,yes,no,no,no,yes,3,no,furnished
2,12250000,9960,3,2,2,yes,no,yes,no,no,2,yes,semi-furnished
3,12215000,7500,4,2,2,yes,no,yes,no,yes,3,yes,furnished
4,11410000,7420,4,1,2,yes,yes,yes,no,yes,2,no,furnished


In [28]:
df_housing.describe()

,price,area,bedrooms,bathrooms,stories,parking
count,5.450000e+02,545.000000,545.000000,545.000000,545.000000,545.000000
mean,4.766729e+06,5150.541284,2.965138,1.286239,1.805505,0.693578
std,1.870440e+06,2170.141023,0.738064,0.502470,0.867492,0.861586
min,1.750000e+06,1650.000000,1.000000,1.000000,1.000000,0.000000
25%,3.430000e+06,3600.000000,2.000000,1.000000,1.000000,0.000000
50%,4.340000e+06,4600.000000,3.000000,1.000000,2.000000,0.000000
75%,5.740000e+06,6360.000000,3.000000,2.000000,2.000000,1.000000
max,1.330000e+07,16200.000000,6.000000,4.000000,4.000000,3.000000


In [33]:
df_housing['price'].max()


13300000

In [32]:
df_housing['price'].min()

1750000

In [26]:


# Смотрим текущие типы и потребление памяти до оптимизации
print("Типы данных до оптимизации:")
print(df_housing.dtypes)
print("\nПотребление памяти до оптимизации:")
print(df_housing.memory_usage(deep=True).sum() / 1024, "КБ")

Типы данных до оптимизации:
price                int64
area                 int64
bedrooms             int64
bathrooms            int64
stories              int64
mainroad            object
guestroom           object
basement            object
hotwaterheating     object
airconditioning     object
parking              int64
prefarea            object
furnishingstatus    object
dtype: object

Потребление памяти до оптимизации:
221.91796875 КБ


1. ЦЕЛОЧИСЛЕННЫЕ КОЛОНКИ (price, area, bedrooms, bathrooms, stories, parking):
   Все 6 колонок были int64 (8 байт). Предлагаю использовать следующие типы:
   - price  → uint32 (4 байта): диапазон 1.75–13.3 млн, uint32 покрывает до 4.3 млрд
   - area   → uint16 (2 байта): диапазон 1,650–16,200, uint16 покрывает до 65,535
   - bedrooms, bathrooms, stories, parking → uint8 (1 байт): значения 0–6
   
   Все значения неотрицательны, поэтому использованы беззнаковые типы (uint).

2. КАТЕГОРИАЛЬНЫЕ КОЛОНКИ (7 колонок типа object):
   - 6 бинарных (yes/no): mainroad, guestroom, basement, hotwaterheating,
     airconditioning, prefarea
   - 1 с 3 значениями (furnished/semi-furnished/unfurnished): furnishingstatus
   
   Предлагаю тип category, который хранит словарь уникальных значений
   + целочисленные коды (1 байт на элемент).



In [37]:

# Диапазон: 1,750,000 – 13,300,000
# uint32 покрывает 0 – 4,294,967,295.
df_housing["price"] = df_housing["price"].astype("uint32")

# Диапазон: 1,650 – 16,200
# uint16 покрывает 0 – 65,535.
df_housing["area"] = df_housing["area"].astype("uint16")


# Диапазон: 1 – 6
# uint8 покрывает 0 – 255.
df_housing["bedrooms"] = df_housing["bedrooms"].astype("uint8")


# Диапазон: 1 – 4
# Аналогично bedrooms: uint8 с большим запасом.

df_housing["bathrooms"] = df_housing["bathrooms"].astype("uint8")


# Диапазон: 1 – 4
# Этажность — малое неотрицательное число.
df_housing["stories"] = df_housing["stories"].astype("uint8")


# Диапазон: 0 – 3
# Количество парковочных мест — малое неотрицательное целое.
df_housing["parking"] = df_housing["parking"].astype("uint8")

# --- mainroad, guestroom, basement, hotwaterheating,
#     airconditioning, prefarea: object → category ---
# Каждая колонка содержит 2 уникальных значения: "yes"/"no".
# category сохраняет исходные строковые метки для удобства.
for col in ["mainroad", "guestroom", "basement", "hotwaterheating",
            "airconditioning", "prefarea"]:
    df_housing[col] = df_housing[col].astype("category")

# --- furnishingstatus: object → category ---
# 3 уникальных значения: "furnished", "semi-furnished", "unfurnished".
# Category хранит 3 элемента в словаре + 1 байт на код для каждой строки.

df_housing["furnishingstatus"] = df_housing["furnishingstatus"].astype("category")





In [38]:
# Смотрим текущие типы и потребление памяти после оптимизации
print("Типы данных после оптимизации:")
print(df_housing.dtypes)
print("\nПотребление памяти после оптимизации:")
print(df_housing.memory_usage(deep=True).sum() / 1024, "КБ")

Типы данных после оптимизации:
price                 uint32
area                  uint16
bedrooms               uint8
bathrooms              uint8
stories                uint8
mainroad            category
guestroom           category
basement            category
hotwaterheating     category
airconditioning     category
parking                uint8
prefarea            category
furnishingstatus    category
dtype: object

Потребление памяти после оптимизации:
10.6953125 КБ


Потребление памяти до оптимизации:
221.91796875 КБ
Потребление памяти после оптимизации:
10.6953125 КБ

Экономия более, чем в 20 раз.